rfc/rfc-91/rfc-91.md

```diff
@@ -0,0 +1,145 @@
+
+## Implementation
+
+This design implements a leader election algorithm for Apache Hudi using a single lock file per table stored in cloud storage. Each table’s lock is represented by a JSON file with the following fields:
+- owner: A unique UUID identifying the lock provider instance.
+- expiration: A UTC timestamp indicating when the lock expires.
+- expired: A boolean flag marking the lock as released.
+
+Example lock file path: `s3://bucket/locks/<cleaned-table-base-path>-<short-sha>.json`
+
+Each `LockProvider` must implement `tryLock()` and `unlock()` however we also need to do our own lock renewal, therefore this implementation also has `renewLock()`. The implementation will import a service using reflection which writes to S3/GCS/Azure based on the provided location to write the locks. This ensures the main logic for conditional writes is shared regardless of the underlying storage.
+
+`tryLock()`
+- No Existing Lock: If the lock file doesn’t exist, a new lock file is created with the current instance’s details using a conditional write that only succeeds if the file is absent.
+- Existing Lock – Not Expired: If a valid (non-expired) lock exists, the process refrains from taking the lock.
+- Existing Lock – Expired: If the lock file exists but is expired, a new lock file is conditionally written. This write uses a precondition based on the current file’s unique tag from cloud storage to ensure that no other process has updated it in the meantime.
+
+`renewLock()`
+- Purpose: Periodically extend the lock’s expiration (the heartbeat).
+- Mechanism: Update the lock file’s expiration using a conditional write that verifies the unique tag from the current lock state. If the tag does not match, the renewal fails, indicating that the lock has been lost.
+
+`unlock()`
+- Purpose: Safely release the lock.
+- Mechanism: Update the lock file to mark it as expired. This update is performed with a conditional write that ensures the operation is only executed if the file’s unique tag still matches the one held by the lock owner.
+
+### Heartbeat Manager
+
+Once a lock is acquired, a dedicated heartbeat task periodically calls renewLock() (typically every 30 seconds) to extend the expiration. This ensures the lock remains valid as long as the owning process (thread) is active. The heartbeat manager oversees this process, ensuring no other updates occur concurrently on the lock file. Each lock provider has one heartbeat manager with a single executor thread.
+
+### Edge cases
+- If the thread which acquired the lock dies, we stop the heartbeat.
+- If the renewal fails past the expiration, we log an error, and stop the heartbeat. Other Hudi lock provider implementations are susceptible to this behavior. If a writer somehow loses access to Zookeeper, there is no way to tell the writer to exit gracefully.
+- If we are unable to start the heartbeat (renewal) we throw exception.
```

Comment:
Will the lock holder perceive the exception in this case, or continue to finish the writing/txn

Reply:
If we cannot start the heartbeat we throw an exception but also release the lock. So the writer will never start a txn.

For renewal failures the writer will continue to finishing the txn